### PR TITLE
CB-10612 Environment service changes for in-place enablement of load balancers

### PR DESCRIPTION
**Pull request by hreeve-cloudera**
Implements the logic to update an environment with load balancers. There are two steps
to the flow:
  1. Updates the saved environment network object.
    - Enables the endpoint gateway if desired.
    - If endpoint gateway subnets are provided, fetchs the metadata for those subnets.
    - Validates the updated network settings.
    - Saves updated network to database.
  2. For each data lake and data hub, initiates the internal update_load_balancers Stack
     service call (currently stubbed out) and waits for the results.

Tested in local CB environment and with unit tests.

Currently based on branch CB-10612-a. Will be rebased on master when that PR is merged in.

### CODE REVIEW

File: environment/src/main/java/com/sequenceiq/environment/environment/flow/EnvironmentReactorFlowManager.java [hunk 1]
```diff
@@ -134,4 +139,22 @@ public void triggerStackConfigUpdatesFlow(Environment environment, String userCr
         eventSender.sendEvent(envStackConfigUpdatesEvent,
             new Event.Headers(getFlowTriggerUsercrn(userCrn)));
     }
+
+    public void triggerLoadBalancerUpdateFlow(EnvironmentDto environmentDto, Environment environment,
+            PublicEndpointAccessGateway endpointAccessGateway, Set<String> subnetIds, String userCrn) {
+        LOGGER.info("Load balancer update flow triggered.");
```
Comment (doktoric) [inline environment/src/main/java/com/sequenceiq/environment/environment/flow/EnvironmentReactorFlowManager.java:145]: please add some context that we will update with which subnetIds

File: environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentLoadBalancerService.java [hunk 2]
```diff
@@ -1,14 +1,60 @@
...
-        // TODO
+        requireNonNull(environmentDto);
+        requireNonNull(environmentLbDto);
+        if (!entitlementService.publicEndpointAccessGatewayEnabled(ThreadBasedUserCrnProvider.getAccountId()) &&
+                PublicEndpointAccessGateway.ENABLED.equals(environmentLbDto.getEndpointAccessGateway())) {
+            throw new BadRequestException("Public endpoint gateway entitlement is not enabled.");
+        } else {
+            if (!entitlementService.datalakeLoadBalancerEnabled(ThreadBasedUserCrnProvider.getAccountId()) &&
+                !entitlementService.publicEndpointAccessGatewayEnabled(ThreadBasedUserCrnProvider.getAccountId())) {
+                throw new BadRequestException("Neither endpoint gateway or data lake load balancer is enabled. Nothing to do.");
```
Comment (doktoric) [inline environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentLoadBalancerService.java:45]: DAta Lake

File: environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentLoadBalancerService.java [hunk 3]
```diff
@@ -1,14 +1,60 @@
...
+            }
+
+            String accountId = Crn.safeFromString(environmentDto.getResourceCrn()).getAccountId();
+            Environment environment = environmentService
+                .findByResourceCrnAndAccountIdAndArchivedIsFalse(environmentDto.getResourceCrn(), accountId).
+                    orElseThrow(() -> new NotFoundException(
+                        String.format("No environment found with crn '%s'", environmentDto.getResourceCrn())));
```
Comment (doktoric) [inline environment/src/main/java/com/sequenceiq/environment/environment/service/EnvironmentLoadBalancerService.java:52]: I think we populate this to the enduser. the crn does not really help then

File: environment/src/main/java/com/sequenceiq/environment/environment/service/LoadBalancerPollerService.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        Set<String> stackNames;
+        if (PublicEndpointAccessGateway.ENABLED.equals(endpointAccessGateway)) {
+            LOGGER.debug("Updating load balancers for endpoint gateway on data lake and data hubs.");
+            stackNames = getDataLakeAndDataHubNames(environmentCrn, environmentName);
+        } else {
+            LOGGER.debug("Updating load balancer for data lake cluster.");
+            stackNames = getAttachedDatalakeClusters(environmentName);
+        }
+
+        LOGGER.debug("Found {} clusters to update for environment {}.", stackNames.size(), environmentName);
```
Comment (doktoric) [inline environment/src/main/java/com/sequenceiq/environment/environment/service/LoadBalancerPollerService.java:75]: found {} Data Lake

File: environment/src/main/java/com/sequenceiq/environment/environment/service/LoadBalancerPollerService.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        if (stackNames.isEmpty()) {
+            LOGGER.debug("No data lake or data hub clusters found for environment.");
+        } else {
+            try {
+                List<FlowIdentifier> failedFlows = waitStackLoadBalancerUpdate(getPollingConfig(), stackNames, environmentName);
+                LOGGER.debug("Stack load balancer update finished.");
+                if (!failedFlows.isEmpty()) {
+                    LOGGER.error("Found failed flows for stack load balancer updates: " + failedFlows);
+                    throw new UpdateFailedException("Stack update flows failed: " + failedFlows);
```
Comment (doktoric) [inline environment/src/main/java/com/sequenceiq/environment/environment/service/LoadBalancerPollerService.java:84]: Stack should be Data Lake or Data Hub

**Comment by doktoric** [comment]
what happens if the load balancer update fails. I mean what the user can do? Start / STop ? Can he also retry it?

**Comment by hreeve-cloudera** [comment]
> what happens if the load balancer update fails. I mean what the user can do? Start / STop ? Can he also retry it?

If the load balancer update ends in a failed state, then the user can retry. Retrying isn't destructive in any way, and won't create new excess resources if the load balancer had already been successfully created in a previous run.

**Comment by doktoric** [comment]
Thanks @hreeve-cloudera  the change looks good. One last question. I think we should cover the loadbalancer update from audit perspective. @topolyai5 can you please help @hreeve-cloudera regarding the additional audit entry?

**Comment by hreeve-cloudera** [comment]
> Thanks @hreeve-cloudera the change looks good. One last question. I think we should cover the loadbalancer update from audit perspective. @topolyai5 can you please help @hreeve-cloudera regarding the additional audit entry?
> Probably telemetry is also required cc @akanto

Thanks Richard. Do you want that to be part of this change, or in a separate PR?

**Comment by topolyai5** [comment]
> Thanks @hreeve-cloudera the change looks good. One last question. I think we should cover the loadbalancer update from audit perspective. @topolyai5 can you please help @hreeve-cloudera regarding the additional audit entry?
> Probably telemetry is also required cc @akanto

Yes, I can. 

@hreeve-cloudera Please add in a separated pr.

**Review by doktoric — APPROVED** [review]

**Comment by hreeve-cloudera** [comment]
@doktoric @topolyai5 Created JIRA for telemetry and audit updates here: https://jira.cloudera.com/browse/CB-11321

**Pull request merged**